crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -877,6 +881,88 @@ impl<'db, 'ast> NarrowingConstraintsBuilder<'db, 'ast> {
             .tuple_windows::<(&ruff_python_ast::Expr, &ruff_python_ast::Expr)>();
         let mut constraints = NarrowingConstraints::default();
 
+        // Narrow tagged unions of `TypedDict`s with `Literal` keys, for example:
+        //
+        //     class Foo(TypedDict):
+        //         tag: Literal["foo"]
+        //     class Bar(TypedDict):
+        //         tag: Literal["bar"]
+        //     def _(union: Foo | Bar):
+        //         if union["tag"] == "foo":
+        //             reveal_type(union)  # Foo
+        //
+        // Importantly, `my_typeddict_union["tag"]` isn't the place we're going to constraint.
+        // Instead, we're going to constrain `my_typeddict_union` itself.
+        if matches!(&**ops, [ast::CmpOp::Eq | ast::CmpOp::NotEq])
+            && let ast::Expr::Subscript(subscript) = &**left
+            && let lhs_value_type = inference.expression_type(&*subscript.value)
+            // Checking for `TypedDict`s up front isn't strictly necessary, since the intersection
+            // we're going to build is compatible with non-`TypedDict` types, but we don't want to
+            // do the work to build it and intersect it (or for that matter, let the user see it)
+            // in the common case where there are no `TypedDict`s.
+            && is_typeddict_or_union_with_typeddicts(self.db, lhs_value_type)
+            && let Some(subscript_place_expr) = place_expr(&subscript.value)
+            && let Type::StringLiteral(key_literal) = inference.expression_type(&*subscript.slice)
+            && let rhs_type = inference.expression_type(&comparators[0])
+            && matches!(rhs_type, Type::StringLiteral(_) | Type::IntLiteral(_))
```

Comment:
is there any reason not to include `BytesLiteral` types here (and in `all_matching_typeddict_fields_have_literal_types` lower down)? I guess it's unlikely that someone would actually use a `BytesLiteral` tag, but if someone were to, they should behave just the same way as string-literal and int-literal tags, so I don't see any disadvantage to including them here